---
title: "Phylogenetic Simulation"
date: "11/14/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache=TRUE,message=FALSE)
```

## Aim

The aim is to write code to simulate DNA sequence data on a tree given a sequence length $n$ and a nucleotide substitution model parameterized by a $4 \times 4$ time-reversible infinitesimal rate matrix $Q$ parameters,
which itself is parameterized by simplex vectors $\pi \in \Delta_4$ and $\rho \in \Delta_6$.
We begin with code to simulate sequence data on a single edge and then generalize to trees.

#### Parameters

The matrix $Q$ is parameterized by stationary distribution $\pi = \{\pi_i\}$ for $i=1,\ldots,4$ with the constraint $\sum_{i=1}^4 \pi_i = 1$ and rate vector $\rho = \{\rho_{ij}\}$ for $1 \le i < j \le 4$ with the constraint
$\sum_{i=1}^3\sum_{j=i+1}^4 \rho_{ij} = 1$.
We equate the indices $1\ldots,4$ with the DNA bases $\{A,C,G,T\}$.
Furthermore, for $i>j$, let $\rho_{ij} = \rho_{ji}$.
We define
$$
Q(\pi,\rho)_{ij} = \left\{\begin{array}{ll}
\frac{\rho_{ij}}{2\pi_i} & \text{if $i \neq j$} \\
\frac{\sum_{k \neq i}\rho_{ik}}{2\pi_i} & \text{if $i = j$}
\end{array}\right.
$$
In this parameterization,
the vector $\pi$ is the stationary distribution
of the continuous-time Markov chain governed by $Q$ and $\rho_{ij}$ represents the long-run proportion of transitions between states $i$ and $j$ in either direction.

#### Matrix Decomposition

The matrix $Q$ has the decomposition $Q = V \Lambda V^{-1}$
where $V$ is a matrix of eigenvalues of $Q$,
$\Lambda$ is a diagonal matrix of the eigenvalues $\lambda$ of $Q$ (incorresponding order)
and $V^{-1}$ is the matrix inverse of $V$.

#### Transition Matrix

The probability transition matrix over a time $t$ is given by the matrix
$$
P(t) = \mathrm{e}^Qt = V \mathrm{e}^{\Lambda t}V^{-1}
$$

## Functions

Here are the functions needed to simulate data over an edge.

```{r simulate-functions}
## Make matrix Q from pi and rho
make_Q = function(pi,rho)
{
  Q = matrix(0,4,4)
  Q[row(Q) > col(Q)] = rho
  Q = Q + t(Q)
  Q = diag(1/pi) %*% Q
  Q = Q/2
  diag(Q) = -apply(Q,1,sum)
  return ( Q )
}

## Build Q and qlist from pi and rho
build_Q = function(pi,rho)
{
  DNA = c('a','c','g','t')
  Q = make_Q(pi,rho)
  e = eigen(Q)
  lambda = e$values
  V = e$vectors
  Vinv = solve(V)
  dimnames(pi) = NULL
  names(pi) = DNA
  names(rho) = c('ac','ag','at','cg','ct','gt')
  rownames(Q) = DNA
  colnames(Q) = DNA
  return ( list(V=V,Vinv=Vinv,lambda=lambda,Q=Q,pi=pi,rho=rho) )
}

## Make transition matrix P
make_P = function(qlist,t)
{
  DNA = c('a','c','g','t')
  P = qlist$V %*% diag(exp(qlist$lambda*t)) %*% qlist$Vinv
  rownames(P) = DNA
  colnames(P) = DNA
  return ( P )
}

## Generate stationary sequence
generate_stationary = function(n,pi)
{
  return ( sample(1:4,n,pi,replace=TRUE) )
}

## Generate child sequence given parent
## x is a vector with values from 1:4
## P is a 4 by 4 matrix with row sums equal to 1
##  and all nonnegative values
generate_sequence = function(x,P)
{
  y = numeric(length(x))
  bases = 1:4
  for ( b in bases )
  {
    nb = sum(x==b)
    if ( nb > 0 )
    {
      y[x==b] = sample(bases,nb,prob=P[b,],replace=TRUE)
    }
  }
  return ( y )
}

DNA = c('a','c','g','t')
```

### Testing the Sequence Generation Code

```{r test-seq-gen}
library(ape)
### Read in sequence data for realistic pi and rho
sg = read.dna("sheep-giraffe.fasta",
              as.character=TRUE, format="fasta")
n = ncol(sg)
# p is stationary distribution
p = table(sg) / sum(table(sg))
tab = table(sg[1,],sg[2,])
# Transition rate
rho = (tab + t(tab))[row(tab) > col(tab)]
rho = rho/sum(rho)
## Make Q
qlist = build_Q(p,rho)
## Let t = 0.15
P = make_P(qlist,0.15)
# Generate child from parent
x = generate_stationary(n,qlist$pi)
y = generate_sequence(x,P)
```

### Example Tree

```{r example-tree}
##library(ggtree)
library(ape)
tree_string = "((rat:2,mouse:2):1,(horse:3,(cow:2,pig:2):1):3);"
tree = read.tree(text=tree_string)
plot(tree)
```

### Simulate Tree

```{r simulate-tree}
generate_sequence_tree = function(tree,n,qlist)
{
  ## number of nodes
  num_leaves = length(tree$tip.label)
  num_internal = tree$Nnode
  ## root has the smallest number among the internal nodes
  root_node = num_leaves+1
  ## container for all of the sequences
  seq = matrix(0,num_leaves+num_internal,n)
  ## generate data at the root
  seq[root_node,] = root_sequence = generate_stationary(n,qlist$pi)
  ## traverse through the tree and generate sequences at each child
  for ( i in 1:nrow(tree$edge) )
  {
    P = make_P(qlist,tree$edge.length[i])
    parent = tree$edge[i,1]
    child = tree$edge[i,2]
    seq[child,] = generate_sequence(seq[parent,],P)
  }
  dat = seq[1:num_leaves,]
  rownames(dat) = tree$tip.label
  return ( dat )
}
generate_sequence_tree(tree,20,qlist)

## turn a matrix of values from 1 to 4 into a character matrix
##  with the DNA letters
make_dna = function(x)
{
  DNA = c('a','c','g','t')
  out = matrix(DNA[x],nrow(x),ncol(x))
  rownames(out) = rownames(x)
  return ( out )
}
```

### Test

```{r sim-tree}
sim001 = generate_sequence_tree(tree,n,qlist)
make_dna(sim001)[,1:20]
write.dna(make_dna(sim001),file="sim001.fasta",format="fasta",nbcol=-1,colsep="")
```